crates/history/src/browser.rs

```diff
@@ -0,0 +1,385 @@
+
+        self.notify_callbacks();
+        Ok(())
+    }
+
+    #[cfg(feature = "query")]
+    fn push_with_query<'a, Q>(&self, route: impl Into<Cow<'a, str>>, query: Q) -> HistoryResult<()>
+    where
+        Q: Serialize,
+    {
+        let url = route.into();
+        let query = serde_urlencoded::to_string(query)?;
+        self.inner
+            .push_state_with_url(&JsValue::NULL, "", Some(&format!("{}?{}", url, query)))
+            .expect_throw("failed to push history.");
+
+        self.notify_callbacks();
+        Ok(())
+    }
+    #[cfg(feature = "query")]
+    fn replace_with_query<'a, Q>(
+        &self,
+        route: impl Into<Cow<'a, str>>,
+        query: Q,
+    ) -> HistoryResult<()>
+    where
+        Q: Serialize,
+    {
+        let url = route.into();
+        let query = serde_urlencoded::to_string(query)?;
+        self.inner
+            .replace_state_with_url(&JsValue::NULL, "", Some(&format!("{}?{}", url, query)))
+            .expect_throw("failed to replace history.");
+
+        self.notify_callbacks();
+        Ok(())
+    }
+
+    #[cfg(all(feature = "query", feature = "state"))]
+    fn push_with_query_and_state<'a, Q, T>(
+        &self,
+        route: impl Into<Cow<'a, str>>,
+        query: Q,
+        state: T,
+    ) -> HistoryResult<()>
+    where
+        Q: Serialize,
+        T: Serialize + 'static,
+    {
+        let url = route.into();
+        let query = serde_urlencoded::to_string(query)?;
+        let state = serde_wasm_bindgen::to_value(&state)?;
+        self.inner
+            .push_state_with_url(&state, "", Some(&format!("{}?{}", url, query)))
+            .expect_throw("failed to push history.");
+
+        self.notify_callbacks();
+        Ok(())
+    }
+
+    #[cfg(all(feature = "query", feature = "state"))]
+    fn replace_with_query_and_state<'a, Q, T>(
+        &self,
+        route: impl Into<Cow<'a, str>>,
+        query: Q,
+        state: T,
+    ) -> HistoryResult<()>
+    where
+        Q: Serialize,
+        T: Serialize + 'static,
+    {
+        let url = route.into();
+        let query = serde_urlencoded::to_string(query)?;
+        let state = serde_wasm_bindgen::to_value(&state)?;
+        self.inner
+            .replace_state_with_url(&state, "", Some(&format!("{}?{}", url, query)))
+            .expect_throw("failed to replace history.");
+
+        self.notify_callbacks();
+        Ok(())
+    }
+
+    fn listen<CB>(&self, callback: CB) -> HistoryListener
+    where
+        CB: Fn() + 'static,
+    {
+        // Callbacks do not receive a copy of [`History`] to prevent reference cycle.
+        let cb = Rc::new(callback) as Rc<dyn Fn()>;
+
+        self.callbacks.borrow_mut().push(Rc::downgrade(&cb));
+
+        HistoryListener { _listener: cb }
+    }
+
+    fn location(&self) -> Self::Location {
+        BrowserLocation::new(self.clone())
+    }
+}
+
+impl Default for BrowserHistory {
+    fn default() -> Self {
```

Comment:
I saw your messages on the Yew discord advising against directly calling it. Should this be `pub(crate)` function instead of `Default` implementation? That way, it won't be part of the public API

Reply:
I didn’t recommend it due to it is not a good practice inside components and users should use hooks instead. Not that default implementation itself has any issue.

I think an updated implementation will no longer expose History interface via hooks, so I am fine with default implementation.

In addition, it trips up clippy if you have a new method, but no default implementation.